### ISSUE TITLE
[Feature Request] additional auto-type targets with support for wildcards

### ISSUE THREAD

**Issue by nopivnick** [keeweb/keeweb#433]
KeeWeb's auto-type functionality is fairly literal and somewhat rigid when it comes to use of the url field as a target for triggering auto-type.

login prompts don't always reside at the root of the domain, may reside at a different subdomain, or some other variation thereof.

the ability to enter more than one variation of auto-type target with wildcards (or at least recognize those additional targets created with KeePass) would be quite helpful.

**Comment by antelle** [comment]
This is planned for future.
For now, you can add additional url to custom field.

**Comment by Narvey** [comment]
Is what is planned for future just the aspect of how it detects the target window? Or will there also be the ability to actually change __what__ gets auto-typed like issue #434 mentions? I'm getting sick of having to go back to Windows to adjust auto-type sequences...

**Comment by antelle** [comment]
@Narvey, you can change what's get auto-typed, in auto-type sequence.

**Comment by antelle** [comment]
Per-window configuration will also include a custom sequence

**Comment by Raikia** [comment]
+1 to this.  Correct me if I'm wrong, but it looks like Auto-Type (at least on Windows) matches based on the webpage "title", so putting the URL in the additional field doesn't solve anything.  Also there are no wildcard searches on titles (that I could tell), so that doesn't work either (since titles generally change depending on the page you are on).

**Comment by antelle** [comment]
@Raikia, no, URL matters: in the most browsers on Windows and Mac we are able to get the URL and find entry by it.

**Comment by Raikia** [comment]
Oh, well I stand corrected then!  I guess the detection doesn't work on Vivaldi then :-(

**Comment by antelle** [comment]
No, it doesn't work on Vivaldi, Vivaldi doesn't work well with accessibility API.

**Comment by gynet** [comment]
Also, it happens sometimes when a user has one user credential for many applications / URLs, and each of them may have different auto type behaviors, e.g {username} {enter} {password}, or just {passord} 
It would be nice that keeweb woule be able to handle **one to many** situations.

**Comment by antelle** [comment]
@gynet auto type window config supposed to be one to may, you're right.

**Comment by elbowz** [comment]
I don't understand well how the auto type **matching** is done.

I'm on Linux + Chrome + KeeWeb v1.17.5 and here it seems that match is only based on:
`Window title` (Chrome browser) ~= `keeweb entry title`

I'm trying to add/changing the `Website` field in keeweb but it seems not works. 

Here the log:
```
debugAutoType
1
index.html:1 2021-03-31T11:35:23.041Z [auto-type] Auto type event null
index.html:1 2021-03-31T11:35:23.046Z [auto-type] Getting window info using native helper
index.html:1 2021-03-31T11:35:23.052Z [auto-type] Window info undefined Humble Bundle - Log In - Google Chrome null
```

The matching work only if I put in title (of keeweb entry) one or more words from the set: {Humble, Bundle, Log In, Google, Chrome}.

Are there some wiki/doc pages I have missed?

**Comment by antelle** [comment]
@elbowz on Linux KeeWeb tries to get URL using accessibility API (ATSPI), if this fails, it matches only by window title.

I'm not sure why ATSPI sometimes doesn't work, when I tested it on Ubuntu, it starts working after going through `orca` setup. I have no idea why, but you can try it, maybe this will help in your case as well.

Otherwise it will match the way you discovered it, i.e. if the entry matches title.

**Comment by elbowz** [comment]
Thanks for the detailed answer. 

I have tried to follow the information [here](https://techbase.kde.org/Development/Tutorials/Accessibility/Screen_Reader_Setup), because I'm on KDE and ATSPI is a Gnome project.

I have all package installed: at-spi2-atk, at-spi2-core (already present), python-atspi, orca
check the service at-spi-dbus-bus.service running (a sort of bridge from atspi2dbus)

...but keeweb is not able to get current tab url.

Anyway the accessibility information found is quite old and the accessibility tools are not a lot used...so, some miss-configuration on my PC. If someone with KDE will be able to work this feature, please ping me :)

@antelle is there a way to define multiple "title" match, ie. some keeweb entry field like website but for window title, maybe with wildcard?

**Comment by antelle** [comment]
Right, AT-SPI doesn't seem to work in KDE. Do you know if there's any other way to get browser url? I can implement it instead.

Well, the proper answer to it is using a browser add-on that we should have in KeeWeb, since getting browser url seems to be quite a challenge. This will be especially bad in wayland.

For now, there's no way to customize title matching, that's what this issue is about.

The icon seems to be an issue in Electron, afaik there's no way to customize its title. Also the icon doesn't look great 😢

**Comment by elbowz** [comment]
> Right, AT-SPI doesn't seem to work in KDE. Do you know if there's any other way to get browser url? I can implement it instead.

I'm sorry...I didn't even know of the existence of ATSPI api.

> Well, the proper answer to it is using a browser add-on that we should have in KeeWeb, since getting browser url seems to be quite a challenge. This will be especially bad in wayland.

Yeah! This is a great idea. If an official browser add-on exist, I think you can also add some further features. For what I have found on google should be simple, but you have to implement a web(socket) server in keeweb to allow add-on (<)-> keeweb communication.

Do you think that Gnome + wayland even not works?

> For now, there's no way to customize title matching, that's what this issue is about.

Allow the possibility to add multiple custom fields (eg. `match-title`) to a keeweb entry, like already is done with multiple `Website`, maybe with regex support, could be useful.

> The icon seems to be an issue in Electron, afaik there's no way to customize its title. Also the icon doesn't look great

Not a big trouble :)
I perfectly understand how can be complicated develop a multi-platform application (also using electron), especially on the infinite world of Linux package combination

**Comment by antelle** [comment]
@elbowz  the browser add-on is tracked in #874 

> Do you think that Gnome + wayland even not works?

Not sure I understand the question, but auto-type won't work Wayland, at least now, until we get better protocol support. You can find more about it here: https://github.com/keeweb/keeweb/issues/1246#issuecomment-802939965.

> Allow the possibility to add multiple custom fields 

That's what should be implemented: https://aavtech.site/2019/04/keepass-auto-type-guide/, the kdbx format has support for this, we're just not reading it yet in KeeWeb.

**Comment by elbowz** [comment]
> @elbowz the browser add-on is tracked in #874
> 
> > Do you think that Gnome + wayland even not works?
> 
> Not sure I understand the question, but auto-type won't work in Wayland, at least now, until we get better protocol support. You can find more about it here: [#1246 (comment)](https://github.com/keeweb/keeweb/issues/1246#issuecomment-802939965).

ahh ok, now it's more clear...the trouble is related to `xdotool` (virtual keyboard), not to the ATSPI (current browser url).
I guess with a dedicated browser extension also the fields filling could be solved

> > Allow the possibility to add multiple custom fields
> 
> That's what should be implemented: https://aavtech.site/2019/04/keepass-auto-type-guide/, the kdbx format has support for this, we're just not reading it yet in KeeWeb.

Cool, so auto-type is part of kbdx "standard"...so, I'll wait for news on this feature :)

**Comment by antelle** [comment]
Well, we're not using `xdotool` anymore and call XKB API's instead ([here](https://github.com/antelle/keyboard-auto-type/blob/master/keyboard-auto-type/src/linux/auto-type-linux.cpp) is the implementation if you're interested). They of course don't exist on Wayland, where we can either reply on protocols (`virtual-keyboard` or `input-method`) that are not yet widely supported, or write to `/dev/uinput` that has other issues.

**Comment by elbowz** [comment]
So, you have a native compiled part for each OS (ie. Windows, Linux and Mac)...I really do not envy you ;)

I don't know the implications, issues and requirements behind these chose, but I wonder if the browser add-on with a simple js call `document.createEvent('KeyboardEvent')` could do the same job...equal on all platforms/OS.

Anyway, my compliment for your good and hard work...also in the "customer care"!

I'll wait, without hurry, the new "multi title window match" feature.

Thx!

**Comment by antelle** [comment]
Well, browser is just one app where auto-type is used. People also do it in other apps (for example, remote desktop, terminal, and so on), this way works everywhere. So even when we have a proper browser extension, there's still value in this feature request. Actually, we do support have browser extensions, but in a bit old-fashioned way: extensions compatible with KeePassHttp will work with KeeWeb: https://github.com/keeweb/keeweb/wiki/Browser%20AutoFill. It may be a way to go for you now, it's still working well in Chrome.

Yeah, for browsers an add-on will do the same job. It's not that simple though, for example, Safari didn't support Web Extensions before 2020, now it supports them but the way of communication with apps is different. We need to develop all of this as well or reuse parts of existing extensions like KeePassXC-browser, this is being discussed and explained in detail in the linked issue.

**Comment by elbowz** [comment]
> Well, browser is just one app where auto-type is used. 

ohhhh, it's right! I never used in this way, but it has a more sense now

> People also do it in other apps (for example, remote desktop, terminal, and so on), this way works everywhere. So even when we have a proper browser extension, there's still value in this feature request. Actually, we do support browser extensions, but in a bit old-fashioned way: extensions compatible with KeePassHttp will work with KeeWeb: https://github.com/keeweb/keeweb/wiki/Browser%20AutoFill. It may be a way to go for you now, it's still working well in Chrome.

Thanks...I'll try it.

Happy Easter
